Elly Vendingmachine step3

dely2p opened the pull request:
Vendingmachine step3 입니다.

dely2p committed:
fix main code to use description of Class instance by JK feedback #15

CustomStringConvertible 프로토콜 채택 시 .description으로 접근하지 않아도 instance 접근만으로도 print()할 수 있다는 것을 배울 수 있었다.

dely2p committed:
create VendingMachine struct to make several method #11

VendingMachine 구조체를 추가하여 관련 메소드들을 생성하기 위해 구조를 만들었다.

dely2p committed:
create InventoryBox Class to save input inventory #11

재고를 저장하기 위해서 InventoryBox 클래스를 생성함.

dely2p committed:
fix access control to use Beverage in another Class

Beverage 클래스의 private를 지움.

dely2p committed:
remove typeOfBeverage to use typeOfBeverage of SuperClass #11

typeOfBeverage를 슈퍼클래스에서 정의하고 서브클래스에서는 각각의 클래스의 이름으로 초기화하도록 구현했다.

dely2p committed:
add addInInventory method in VendingMachine

재고추가 시 Class의 meta type으로 dictionary의 key값을 갖도록 만들었다.

dely2p committed:
add listOfDrinksAvailable method in VendingMachine to get List of drinks available at the current price #11

현재 금액으로 구매가능한 음료수 목록을 구하는 메소드를 생성했다.

dely2p committed:
add buyBeverage method to remove Beverage in Inventory #11

음료 구매 메소드를 생성하여 재고 중에서 배열 첫번째 Beverage를 꺼내오는 것을 구현했다.

dely2p committed:
add return coins method #11

현재 투입 금액을 return 해주는 메소드 생성 및 음료 구매 시 투입 금액을 음료 가격만큼 마이너스 해주는 것을 구현했다.

dely2p committed:
add method to return Overall product inventor #11

전체 상품 재고를 출력해주는 메소드를 구현했다.

dely2p committed:
add showPurchaseProductHistory method in VendingMachine #11

구매 이력을 배열에 넣어서 보관한 후 메소드 호출 시 리턴하는 것을 구현했다.

dely2p committed:
add menu in InputView Class #12

자판기 초기메뉴를 만들고 음료수 객체 인스턴스를 추가하고 각각 구분하기 위해 kindOf 속성을 Beverage에서 상속받고 각 SubClass에서 사용하도록 수정하였다.

dely2p committed:
add find purchase available Beverage method #12

구매 가능한 음료수 리스트를 보여주는 메소드를 생성했다.

dely2p committed:
add function to purchase Beverage #12

음료수 구매 시 투입한 동전금액 줄이고, 자판기 음료수 개수 줄이는 메소드를 생성했다.

dely2p committed:
add function to show Out Of Stock Inventory #11

유통기한 지난 음료수를 보여주는 기능을 생성했다.

dely2p committed:
add function to show hotDrink #11

뜨거운 음료를 찾아서 넘겨주는 메소드를 생성했다.

dely2p committed:
add UnitTestVendingMachine to Integrated test #14

전체 코드를 확인할 수 있는 통합 테스트 시나리오를 가지고 동작을 확인하기를 위해 테스트코드를 작성했다.

dely2p committed:
fix property typeOfBeverage in each Class #11

클래스 속성에 접근해서 값을 가져오지않고 클래스의 type으로 출력해주도록 수정했다.

godrm commented on VendingMachine/VendingMachine/InputView.swift:
```diff
@@ -0,0 +1,103 @@
+class InputView {
+    var coins: Int
+    var vendingMachine: VendingMachine
+    var availableBeverage: [ObjectIdentifier:[Beverage]]
```
private 속성으로 바꾸세요

godrm commented on VendingMachine/VendingMachine/InputView.swift:
```diff
@@ -0,0 +1,103 @@
+        for (index, menu) in availableBeverage.enumerated() {
+            print("\(index+1))\(menu.value[0].kindOf) \(menu.value[0].price)(\(menu.value.count)개)")
+        }
+    }
```
OutputView가 따로 없나요?

dely2p replied:
OutputView 클래스를 생성하여 InputView의 print()부분을 분리하는 방법으로 수정했습니다.

godrm commented on VendingMachine/VendingMachine/InventoryBox.swift:
```diff
@@ -0,0 +1,75 @@
+        case invalidBeverage
+    }
+    var box: [ObjectIdentifier:[Beverage]]
+    init() {
```
항상 private 형태로 만들고 메소드로 접근하는 방법이 있는지 우선적으로 고려하세요.

godrm commented on VendingMachine/VendingMachine/InputView.swift:
```diff
@@ -0,0 +1,103 @@
+            printDoingMenu()
+            getMenuInput()
+        }while isContinue
+    }
```
이건 InputView가 자판기 전체를 관리하는 듯한 느낌인데요?

godrm reviewed (commented):
구조를 조금만 개선해보세요.

dely2p committed:
fix property accessible control to private in InputView #16

InputView 클래스의 프로퍼티를 private로 수정했다.

dely2p committed:
fix repeat-while() sentence structure from usingVendingMachine() method in InputView to main #17

InputView에서 while문으로 메뉴를 출력하고 자판기의 동작을 작동하도록 했던 구문을 main으로 옮겨서 분리했다.

dely2p committed:
add OutputView Class to separate output syntax #18

OutputView 클래스를 생성해서InputView에서 print로 출력했던 것들을 Output에서 동작하도록 수정했다.

dely2p committed:
add overallProductInventory in InventoryBox to get box private property #19

InventoryBox에 있는 box property를 private로 바꾸고 다른 클래스에서 이 property에 접근하기 위해 overallProductInventory method()를 생성했다.

dely2p committed:
fix accessible control in VendingMachine, PurchaseProductHistory Class #19

VendingMachine, PurchaseProductHistory 클래스의 property를 private로 수정했다.

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -11,23 +11,34 @@ import Foundation
+            inputView.menuOfAddAmount()
+        }
+        outputView.printDoingMenu()
+        inputView.getMenuInput()
```
객체 관계가 이상하네요. InputView, OutputView와 VendingMachine은 서로 의존하도록 작성하면 안됩니다. 사다리 게임을 생각해보세요.
InputView에서 메뉴를 입력하면 해당 메뉴에 따라서 VendingMachine에 있는 특정 메소드를 호출해서 동작하도록 작성해야 합니다.
그리고 그 동작에 대한 처리 결과로 음료수 목록을 OutputView에 전달하면 출력하는 형태가 되어야 할 것 같습니다.
main 부분을 다시 생각해보세요.

dely2p replied:
OutputView와 VendingMachine을 서로 의존하지 않도록 수정했습니다.

dely2p committed:
Improving the relationship between objects #20

클래스 객체 간의 의존성을 줄이기 위해 main에서 객체를 생성하여 호출하고, OutputView에서 VendingMachine의 값을 가져와서 출력하도록 개선했다.

dely2p committed:
remove unused BeverageMenu Class #20

사용하지 않는 BeverageMenu Class를 삭제했다.

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -26,18 +27,29 @@ func run() {
-    let outputView = OutputView()
+    // 메뉴 출력 및 사용자 입력
+    let inputView = InputView()
+    let outputView = OutputView(vendingMachine: vendingMachine)
```
객체 관계 개선에서 InputView는 좋은데
여전히 OutputView가 VendingMachine에 의존하고 있네요.
이 관계를 끊어보세요. 사다리게임이나 JSON분석기 미션을 떠올려보세요.

dely2p replied:
OutputView와 VendingMachine을 서로 의존하지 않도록 수정했습니다. :)

dely2p committed:
fix struct to reduce dependency in OutputView Class #21

OutputView 안에 의존하고 있는 VendingMachine을 main으로 분리한다.

dely2p committed:
fix white space by codacy #21

codacy에서 발견한 공백 및 띄어쓰기를 수정했다.

godrm commented on VendingMachine/VendingMachine/OutputView.swift:
```diff
@@ -30,13 +27,12 @@ class OutputView {
+    func printCurrentCoins() -> Int {
         print("\n현재 투입한 금액이 \(self.coins)원입니다. 다음과 같은 음료가 있습니다.")
-        vendingMachine.putCoins(coins: coins)
+        return self.coins
```
어라 OutputView가 왜 coins를 갖고 있는건가요? 
어차피 vendingMachine 내부에서 계산하지 않나요?

dely2p replied:
vendingMachine에서 계산하긴 하지만 OutputView 내의 음료를 구매할 때 어떤 음료를 구입하는 지 출력과 함께 음료수의 가격을 차감할 때(printPurchaseBeverage() method)도 필요하고, 현재 남아있는 금액을 보여주는 부분(printCurrentCoins() method)에서도 필요해서 OutputView 클래스 내에 전역변수로 두었습니다.

원래 vendingMachine의 coins 속성을 private로 두지 않았을 때는 coins값이 필요할 때 직접 coins에 접근했었는데 수정하면서 OutputView에 따로 coins를 만들어 사용하게 되었습니다.
vendingMachine에 getCoins() method를 두어서 coins의 값을 가져오게 하는 것이 더 좋을까요??

dely2p replied:
VendingMachine private(set) coins 프로퍼티에 접근하여 금액을 가져오도록 수정하여 OutputView에서는 따로 coins 프로퍼티를 사용하지 않는 방법으로 개선했습니다.

dely2p committed:
fix coin property from private to private(set) #22

OutputView의 coins를 없애고 VendingMachine에서 직접 접근하도록 수정했다.

godrm commented:
다음 단계로 갑시다~
